tests/test_app_client_signer_sender.py

```diff
@@ -0,0 +1,148 @@
+        (only_default_signer, signer_with_address_override_sender, signer_with_address_override_sender),
+        (only_default_signer, only_signer_with_address, signer_with_address_and_associated_sender),
+        (only_signer_with_address, override_signer_sender, override_signer_sender),
+        (only_signer_with_address, only_override_sender, signer_with_address_override_sender),
+        (only_signer_with_address, no_signer_sender, signer_with_address_and_associated_sender),
+        (only_signer_with_address, signer_with_address_override_sender, signer_with_address_override_sender),
+        (only_signer_with_address, only_signer_with_address, signer_with_address_and_associated_sender),
+        (only_default_sender, override_signer_sender, override_signer_sender),
+        (only_default_sender, only_override_signer, override_signer_default_sender),
+        (only_default_sender, signer_with_address_override_sender, signer_with_address_override_sender),
+        (only_default_sender, only_signer_with_address, signer_with_address_default_sender),
+        (no_signer_sender, override_signer_sender, override_signer_sender),
+        (no_signer_sender, signer_with_address_override_sender, signer_with_address_override_sender),
+        (no_signer_sender, only_signer_with_address, signer_with_address_and_associated_sender),
+    ],
+    ids=use_globals_name,
+)
+def test_resolve_signer_sender(
+    algod_client: AlgodClient,
+    app_spec: ApplicationSpecification,
+    created_with: SignerSender,
+    called_with: SignerSender,
+    expected: SignerSender,
+) -> None:
+    app_client = ApplicationClient(algod_client, app_spec, signer=created_with.signer, sender=created_with.sender)
+
+    signer, sender = app_client.resolve_signer_sender(called_with.signer, called_with.sender)
+
+    assert signer == expected.signer
+    assert sender == expected.sender
+
+
+NoSenderError = "No sender provided"
+NoSignerError = "No signer provided"
+
+
+@pytest.mark.parametrize(
+    ("created_with", "called_with", "expected"),
+    [
+        (only_default_signer, only_override_signer, NoSenderError),
+        (only_default_signer, no_signer_sender, NoSenderError),
+        (only_signer_with_address, only_override_signer, NoSenderError),
+        (only_default_sender, only_override_sender, NoSignerError),
+        (only_default_sender, no_signer_sender, NoSignerError),
+        (no_signer_sender, only_override_sender, NoSignerError),
+        (no_signer_sender, only_override_signer, NoSenderError),
+        (no_signer_sender, no_signer_sender, NoSignerError),
+    ],
+    ids=use_globals_name,
+)
+def test_resolve_signer_sender_failure(
+    algod_client: AlgodClient,
+    app_spec: ApplicationSpecification,
+    created_with: SignerSender,
+    called_with: SignerSender,
+    expected: str,
+) -> None:
+    app_client = ApplicationClient(algod_client, app_spec, signer=created_with.signer, sender=created_with.sender)
+
+    with pytest.raises(Exception) as ex:
```

Comment:
this should use `match` btw, instead of the assert below?

Reply:
Done